edsnlp/tune.py

```diff
@@ -0,0 +1,400 @@
+) -> None:
+    """
+    Perform two-phase hyperparameter tuning using Optuna.
+
+    This method executes a two-phase tuning strategy. In the first phase, all specified
+    hyperparameters are tuned. Based on their computed importance, only the most
+    important hyperparameters (top 50% by importance) are selected for fine-tuning
+    in the second phase, while the less important hyperparameters are frozen to their
+    best values from phase one.
+
+    Parameters:
+    -----------
+    config : dict
+        The configuration dictionary for the model and training process.
+
+    hyperparameters : dict
+        A dictionary specifying the hyperparameters to tune. The keys are the parameter
+        names, and the values are dictionaries containing the following fields:
+        - "path": List[str] representing the path to the parameter in `config`.
+        - "type": The type of parameter ("float", "int", "categorical").
+        - "low": (optional) Lower bound for numerical parameters.
+        - "high": (optional) Upper bound for numerical parameters.
+        - "step": (optional) Step size for numerical parameters.
+        - "log": (optional) Whether to sample numerical parameters on a log scale.
+        - "choices": (optional) List of values for categorical parameters.
+
+    output_dir : str
+        Directory where tuning results, visualizations, and best parameters will
+        be saved.
+
+    n_trials : int
+        The total number of trials to execute across both tuning phases.
+        This number will be split between the two phases, with approximately half
+        of the trials assigned to each phase.
+    """
+    n_trials_2 = n_trials // 2
+    n_trials_1 = n_trials - n_trials_2
+
+    logger.info(f"Phase 1: Tuning all hyperparameters ({n_trials_1} trials).")
+    study = optimize(config, hyperparameters, n_trials_1)
+    best_params, importances = process_results(study, f"{output_dir}/phase_1")
+
+    hyperparameters_to_keep = list(importances.keys())[
+        : math.ceil(len(importances) / 2)
+    ]
+
+    hyperparameters_phase_2 = {
+        key: value
+        for key, value in hyperparameters.items()
+        if any(param in key for param in hyperparameters_to_keep)
+    }
+    hyperparameters_frozen = {
+        key: value
+        for key, value in hyperparameters.items()
+        if not any(param in key for param in hyperparameters_to_keep)
+    }
+
+    _, updated_config = update_config(
+        config, hyperparameters_frozen, values=best_params
+    )
+
+    logger.info(
+        f"Phase 2: Tuning {hyperparameters_to_keep} hyperparameters "
+        f"({n_trials_2} trials). Other hyperparameters frozen to best values."
+    )
+    study = optimize(updated_config, hyperparameters_phase_2, n_trials_2, study)
+    process_results(study, f"{output_dir}/phase_2")
+
+
+@app.command(name="tuning", registry=registry)
+def tune(
+    *,
+    config_meta: Dict,
+    hyperparameters: Dict[str, Dict],
```

Comment:
Since we are already using pydantic, could you add a `class HyperparameterConfig(pydantic.BaseModel)` to type the expected dict. For example, a user might make a mistake and pass
```
my_param:
   ...
   logarithmic: true
   # instead of log: true
```
which would silently fail if we don't do type checking